### DISCUSSION TITLE
Windows installation

### DISCUSSION THREAD

**Discussion by soizicg** [Parcels-code/Parcels#1553] [General]
Hi all, I have been successfully installing parcels on linux and older versions on windows, however I have been having issues installing the last version on windows.
When I activate the parcels environment all is well, however if I try to use the parcel module I always get an error (see following text showing the output after parcel installation and trying to import parcel in the python terminal). 
Also I realised that when I did the "pip install" the output showed "Successfully installed parcels-0.0.0" so it doesn't give the right parcels version.

I have tried to reinstall parcels a few times and I always get the same issue. Any help would be great, thanks.

> (parcels) PS C:\Users\szg23nnj\Documents\oceanparcels\parcels-master> pip install --no-build-isolation --no-deps -e .
> Obtaining file:///C:/Users/szg23nnj/Documents/oceanparcels/parcels-master
>   Checking if build backend supports build_editable ... done
>   Preparing editable metadata (pyproject.toml) ... done
> Building wheels for collected packages: parcels
>   Building editable for parcels (pyproject.toml) ... done
>   Created wheel for parcels: filename=parcels-0.0.0-0.editable-py3-none-any.whl size=5603 sha256=44b365a317bdd8e87cc94e6d69f3010a432b80d5b7aa5855ec0a3c3c5abd8ebf
>   Stored in directory: C:\Users\szg23nnj\AppData\Local\Temp\pip-ephem-wheel-cache-6ge_gral\wheels\35\55\53\4b171eb195665592a828277d52805015af5382ebf95f09c83d
> Successfully built parcels
> Installing collected packages: parcels
> Successfully installed parcels-0.0.0
> (parcels) PS C:\Users\szg23nnj\Documents\oceanparcels\parcels-master> python
> Python 3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:20:11) [MSC v.1938 64 bit (AMD64)] on win32
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import parcels
> Traceback (most recent call last):
>   File "C:\Users\szg23nnj\Documents\oceanparcels\parcels-master\parcels\_version.py", line 5, in <module>
>     version = subprocess.check_output(['git', '-C', os.path.dirname(__file__), 'describe', '--tags'], stderr=subprocess.PIPE).decode('ascii').strip()
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\szg23nnj\Documents\micromamba\envs\parcels\Lib\subprocess.py", line 466, in check_output
>     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\szg23nnj\Documents\micromamba\envs\parcels\Lib\subprocess.py", line 571, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['git', '-C', 'C:\\Users\\szg23nnj\\Documents\\oceanparcels\\parcels-master\\parcels', 'describe', '--tags']' returned non-zero exit status 128.
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "C:\Users\szg23nnj\Documents\oceanparcels\parcels-master\parcels\__init__.py", line 1, in <module>
>     from ._version import version
>   File "C:\Users\szg23nnj\Documents\oceanparcels\parcels-master\parcels\_version.py", line 7, in <module>
>     from parcels._version_setup import version as version  # noqa
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ModuleNotFoundError: No module named 'parcels._version_setup'

**Comment by erikvansebille** [comment]
Thanks for reporting, @soizicg. The preferred (and tested) method of installing Parcels on Windows is via Conda, see also the [installation instructions](https://docs.oceanparcels.org/en/latest/installation.html). Can you see if that works for you?

Regardless, the missing `parcels._version_setup` could be because the following code in pyproject.toml is not executed
https://github.com/OceanParcels/parcels/blob/a8aeca841f4d7d4ce0e957c7eb7830b41a881cbb/pyproject.toml#L34-L36
May that be because of any of the options in your `pip install` (i.e. `--no-build-isolation --no-deps -e`)?

While we indeed have a pypi project for Parcels, I don't have enough experience with the pypi workflow to extensively test or maintain it. Help is very welcome!

**Reply by soizicg** [reply]
I have tried with with conda too and it gives me the same error. I have also tried to install older versions (3.0.1 and 3.0.0) and again same error.
Agree it looks like the pip install is ignoring some part of the pyproject.toml, but I am not sure how to fix this. 
If anybody can offer any help that would be great because it is impossible for me to use parcels with this error.

**Reply by erikvansebille** [reply]
Hmm, strange that it also didn't work with conda. Could you copy-paste the output below, so I can have a look and perhaps spot what goes wrong?

**Reply by soizicg** [reply]
Here is the output I get after the pip install using conda:

> (base) C:\Users\szg23nnj\Documents\oceanparcels\parcels-master>conda activate parcels
> (parcels) C:\Users\szg23nnj\Documents\oceanparcels\parcels-master>pip install --no-build-isolation --no-deps -e .
> Obtaining file:///C:/Users/szg23nnj/Documents/oceanparcels/parcels-master
>   Checking if build backend supports build_editable ... done
>   Preparing editable metadata (pyproject.toml) ... done
> Building wheels for collected packages: parcels
>   Building editable for parcels (pyproject.toml) ... done
>   Created wheel for parcels: filename=parcels-0.0.0-0.editable-py3-none-any.whl size=5603 sha256=3b02db02c83448158839f8037059469f06818fdfd0016f5e10f304e8df34065c
>   Stored in directory: C:\Users\szg23nnj\AppData\Local\Temp\pip-ephem-wheel-cache-ejvj860y\wheels\35\55\53\4b171eb195665592a828277d52805015af5382ebf95f09c83d
> Successfully built parcels
> Installing collected packages: parcels
> Successfully installed parcels-0.0.0

And then the same error when trying to load parcels:

> (parcels) C:\Users\szg23nnj\Documents\oceanparcels\parcels-master>python -c "import parcels"
> Traceback (most recent call last):
>   File "C:\Users\szg23nnj\Documents\oceanparcels\parcels-master\parcels\_version.py", line 5, in <module>
>     version = subprocess.check_output(['git', '-C', os.path.dirname(__file__), 'describe', '--tags'], stderr=subprocess.PIPE).decode('ascii').strip()
>               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\szg23nnj\AppData\Local\miniconda3\envs\parcels\Lib\subprocess.py", line 466, in check_output
>     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\szg23nnj\AppData\Local\miniconda3\envs\parcels\Lib\subprocess.py", line 571, in run
>     raise CalledProcessError(retcode, process.args,
> subprocess.CalledProcessError: Command '['git', '-C', 'C:\\Users\\szg23nnj\\Documents\\oceanparcels\\parcels-master\\parcels', 'describe', '--tags']' returned non-zero exit status 128.
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "<string>", line 1, in <module>
>   File "C:\Users\szg23nnj\Documents\oceanparcels\parcels-master\parcels\__init__.py", line 1, in <module>
>     from ._version import version
>   File "C:\Users\szg23nnj\Documents\oceanparcels\parcels-master\parcels\_version.py", line 7, in <module>
>     from parcels._version_setup import version as version  # noqa
>     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ModuleNotFoundError: No module named 'parcels._version_setup'
> 

Thanks for the help.

**Reply by erikvansebille** [reply]
Huh, but after a `conda create -n parcels -c conda-forge parcels` you don't need to _also_ `pip install` it. 
> (parcels) C:\Users\szg23nnj\Documents\oceanparcels\parcels-master>pip install --no-build-isolation --no-deps -e .

Why do you add that line?

**Reply by soizicg** [reply]
Sorry I didn't mention earlier but I need to use the developer installation because I have played a bit with the way parcel output things.
So I have been following this section:
https://docs.oceanparcels.org/en/latest/installation.html#installation-for-developers

**Reply by erikvansebille** [reply]
Ah ok, I see. And what if you simply do `pip install .`, like is done in the Github Action configuration (which does work for Windows)?
https://github.com/OceanParcels/parcels/blob/a8aeca841f4d7d4ce0e957c7eb7830b41a881cbb/.github/actions/install-parcels/action.yml#L30

**Reply by soizicg** [reply]
Good shout, I tried the pip install without any options and now it raised an error, including the following suggestion:

> Make sure you're either building from a fully intact git repository or PyPI tarballs. Most other sources (such as GitHub's tarballs, a git checkout without the .git folder) don't contain the necessary metadata and will not work.
> 
>       For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj

I was using the parcels code coming from the [zip download](https://github.com/OceanParcels/parcels/archive/refs/heads/master.zip), so I didn't use git to get the source code. Once I used the git clone function everything was solved!

Apologies, I should know better than to not closely follow the instructions :)
Thanks a lot.

**Reply by erikvansebille** [reply]
OK, it would be worth to mention on the installation page that this could be a solution for people who want to install the developer version on Windows

Which steps exactly did you take that worked? 

And would you be willing to create a PR with a note at the bottom of https://github.com/OceanParcels/parcels/blob/master/docs/installation.rst? Or do you want me to add such a note?

**Reply by soizicg** [reply]
Actually your "Installation for developers" section is also good for Windows, I just used the steps given there.
My mistake was that I didn't exactly follow the step 2 since I used a zip download of the source code instead of using the "git clone" function.

But do you want me to be highlight in this section that this method is working for both Linux and Windows?

**Reply by erikvansebille** [reply]
Ah ok, in that case let's just leave the installation instructions as is...
